CMakeLists.txt

```diff
@@ -69,12 +70,31 @@ else()
     message(STATUS "Configuring Hyprland in Release with CMake")
 endif()
 
-include_directories(
-  .
-  "subprojects/wlroots/include/"
-  "subprojects/wlroots/build/include/"
-  "subprojects/udis86/"
-  "protocols/")
+set(WLROOTS_SOVERSION 12032)
+
+find_program(MESON_EXE NAMES meson)
+find_program(NINJA_EXE NAMES ninja)
+find_program(SED_EXE NAMES sed)
+ExternalProject_Add(wlroots_build
+    PREFIX wlroots
+    SOURCE_DIR ${CMAKE_SOURCE_DIR}/subprojects/wlroots
+    PATCH_COMMAND ${SED_EXE} -E -i -e "s/(soversion = 12)([^032]|$$)/soversion = ${WLROOTS_SOVERSION}/g" ${CMAKE_SOURCE_DIR}/subprojects/wlroots/meson.build
+    CONFIGURE_COMMAND ${MESON_EXE} setup --prefix ${CMAKE_CURRENT_BINARY_DIR}/wlroots . ${CMAKE_SOURCE_DIR}/subprojects/wlroots
+    BUILD_COMMAND ${NINJA_EXE} -C .
+    INSTALL_COMMAND ${NINJA_EXE} -C . install)
+
+set(wlroots_INCLUDE_DIR ${CMAKE_SOURCE_DIR}/subprojects/wlroots/include/)
+set(wlroots_LIBRARY ${CMAKE_CURRENT_BINARY_DIR}/wlroots/lib/libwlroots.so.${WLROOTS_SOVERSION})
+
+add_library(wlroots SHARED IMPORTED)
+add_dependencies(wlroots wlroots_build)
+set_target_properties(wlroots PROPERTIES
+    IMPORTED_LOCATION ${wlroots_LIBRARY}
+    INTERFACE_INCLUDE_DIRECTORIES ${wlroots_INCLUDE_DIR})
+
+add_subdirectory("subprojects/udis86" udis86)
```

Comment:
udis needs to be built, no? or does this build?

https://github.com/hyprwm/Hyprland/blob/642030f959efe7bd2692bf6a4236d4c1b39ee51a/Makefile#L92

Reply:
It adds udis' cmake project to the main project, and then later its added as a dependency (`target_link_libraries`) which causes it to be built automatically by cmake.